src/chainstate/stacks/index/marf.rs

```diff
@@ -718,384 +741,328 @@ mod test {
-
-        let block_header_1 = BlockHeaderHash::from_bytes(&[0u8; 32]).unwrap();
-        MARF::format(&mut f, &block_header_1).unwrap();
-
-        // make a deep path
-        let path_segments = vec![
-            (vec![], 0),
-            (vec![], 1),
-            (vec![], 2),
-            (vec![], 3),
-            (vec![], 4),
-            (vec![], 5),
-            (vec![], 6),
-            (vec![], 7),
-            (vec![], 8),
-            (vec![], 9),
-            (vec![], 10),
-            (vec![], 11),
-            (vec![], 12),
-            (vec![], 13),
-            (vec![], 14),
-            (vec![], 15),
-            (vec![], 16),
-            (vec![], 17),
-            (vec![], 18),
-            (vec![], 19),
-            (vec![], 20),
-            (vec![], 21),
-            (vec![], 22),
-            (vec![], 23),
-            (vec![], 24),
-            (vec![], 25),
-            (vec![], 26),
-            (vec![], 27),
-            (vec![], 28),
-            (vec![], 29),
-            (vec![], 30),
-            (vec![], 31),
-        ];
-        let path = vec![0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31];
-
-        let (nodes, node_ptrs, hashes) = make_node4_path(&mut f, &path_segments, [31u8; 40].to_vec());
-        dump_trie(&mut f);
-
-        for i in 1..31 {
-            test_debug!("----------------");
-            test_debug!("i = {}", i);
-            test_debug!("----------------");
-
-            // switch to the next block
-            let next_block_header = BlockHeaderHash::from_bytes(&[i as u8; 32]).unwrap();
-
-            // add a leaf at the end of the path
-            let mut next_path = path.clone();
-            next_path[i] = 32;
-            
-            let triepath = TriePath::from_bytes(&next_path[..]).unwrap();
-            let value = TrieLeaf::new(&vec![], &[i as u8; 40].to_vec());
-            
-            test_debug!("----------------");
-            test_debug!("insert");
-            test_debug!("----------------");
-            MARF::insert_leaf(&mut f, &next_block_header, &triepath, &value).unwrap();
-
-            // verify that this leaf exists in _this_ Trie
-            test_debug!("----------------");
-            test_debug!("get");
-            test_debug!("----------------");
-            let read_value = MARF::get_path(&mut f, &next_block_header, &TriePath::from_bytes(&next_path[..]).unwrap()).unwrap().unwrap();
-            assert_eq!(read_value.data.to_vec(), [i as u8; 40].to_vec());
-            assert_eq!(read_value.path, next_path[i+1..].to_vec());
-            assert_eq!(f.get_cur_block(), next_block_header);
-
-            // can get all previous leaves from _this_ Trie
-            for j in 1..(i+1) {
-                test_debug!("----------------");
-                test_debug!("get-prev {} of {}", j, i);
-                test_debug!("----------------");
+        marf_walk_cow_test("/tmp/rust_marf_walk_cow_node4_20", |s| {
```

Comment:
I'll let it pass this time, but it makes me uncomfortable to both drastically change the behavior of consensus-critical code _and_ refactor its unit tests in the same PR.  It would be better to leave the tests as untouched as possible -- i.e. no changes outside of what is necessary to get them to work with new interfaces (i.e. if the refactoring introduces a bug, we might unknowningly let a regression slip by in the consensus-critical code).  In the future, once we're sure the changes to the consensus-critical code are working, we can refactor the unit tests as a follow-up PR.

Reply:
Okay, sounds good. I do believe that we should make an effort to eliminate boiler plate code even in test cases-- boiler plate obscures what the test cases are doing, and makes refactoring of code even harder, since an update to an internal interface would require updates to each test case's boiler plate.

Reply:
I agree with everything you wrote.  All I was suggesting is that test refactoring should be separate PRs from code refactoring, so we can be more secure in the knowledge that the code is working according to the (fixed) tests and the tests are working according to the (fixed) code.